tf publishing

cmower opened the pull request:
# Basic changes

Old method of sensor frame extraction:

1. Retrieve data from polhemus
2. Convert data to `kdl` frame 
3. Convert to `geometry_msgs`
4. Publish `geometry_msgs` to ROS

New method of sensor frame extraction:

1. Retrieve data from polhemus
2. Convert data to `tf`transform
3. Broadcast `tf` transform

(uses `tf2`)

Re `CMakeLists.txt` and `package.xml`, minor changes to suit files in `src` directory. Removed anything that was unnecessary. 

Removed all things related to `goal_pub` and `obs_pub`. This seemed to be specific to past work and now not relevant. 

No changes to `protocol.h` and `liberty.*`.

# Launch file changes

Can now set zenith for hemisphere. 

# Include dir

Removed. Not needed at all for `tf` extraction and broadcasting.

# Readme changes

Added more descriptions and instructions. 

# Tests

All 4 sensors have been tested successfully on lab computer and my dell laptop. Have not tested with 8 sensor polhemus but should work.

wxmerkt commented on CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 cmake_minimum_required(VERSION 2.8.3)
```
``2.8.12`` because of compile options to be specified below

wxmerkt commented on CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 cmake_minimum_required(VERSION 2.8.3)
-project(polhemus_ros)
+project(polhemus)
```
Perhaps consider ``polhemus_driver`` or ``polhemus_ros_driver`` in line with over ROS drivers. This would setup a library ``libpolhemus`` which may be reserved for the original software which is not ROS-bound.

wxmerkt commented on CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
-project(polhemus_ros)
+project(polhemus)
 
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++0x")
```
Replace with ``add_compile_options(-std=c++11)``

wxmerkt commented on CMakeLists.txt:
```diff
@@ -53,39 +53,30 @@ else (USB_LIBRARIES AND USB_INCLUDE_DIRS)
-
 find_package(catkin REQUIRED COMPONENTS
-  geometry_msgs
   tf
```
Switch to tf2

wxmerkt commented on CMakeLists.txt:
```diff
@@ -53,39 +53,30 @@ else (USB_LIBRARIES AND USB_INCLUDE_DIRS)
-  DEPENDS system_lib orocos_kdl eigen
-)
-
+  LIBRARIES polhemus
```
As above, this will block the polhemus name and potentially create conflict.

cmower replied:
@wxmerkt should this be changed to 
```
catkin_package(
  LIBRARIES polhemus_ros_driver
  CATKIN_DEPENDS geometry_msgs tf roscpp
  )
```
?
Note, I've changed the package name to `polhemus_ros_driver`.

wxmerkt replied:
It depends on whether you're also changing the name of the library. I would leave the name of the library as polhemus if it is only the liberty stuff.

wxmerkt commented on launch/start.launch:
```diff
@@ -0,0 +1,39 @@
+  <param name="z_hs" value="-1" type="double"/>
+
+  <!--Setup static transform polhemus_base <-> world-->
+  <node pkg="tf" type="static_transform_publisher" name="polhemus_world_broadcaster" args="0 0 0.54 0 0 3.1415926 polhemus_world polhemus_base 100"/>
```
This should be in application-specific launch files, but good to leave as an example

VladimirIvan replied:
This is a very specific transformation.
Either set everything to default/zero values or rename the file to `example.launch`.

wxmerkt commented on launch/start.launch:
```diff
@@ -0,0 +1,39 @@
+  <node pkg="tf" type="static_transform_publisher" name="polhemus_world_broadcaster" args="0 0 0.54 0 0 3.1415926 polhemus_world polhemus_base 100"/>
+
+  <!--Specify attached stations and those to broadcast-->
+  <param name="station_1_attached" value="true" type="bool"/>
```
FYI: You can specify these via YAML, but okay this way as well.

VladimirIvan commented on CMakeLists.txt:
```diff
@@ -53,39 +53,30 @@ else (USB_LIBRARIES AND USB_INCLUDE_DIRS)
   ${catkin_LIBRARIES}
   ${USB_LIBRARIES}
-)
+  )
```
Please add install directives

cmower replied:
@VladimirIvan sorry, what do you mean?

VladimirIvan replied:
http://wiki.ros.org/catkin/CMakeLists.txt#Optional_Step:_Specifying_Installable_Targets

wxmerkt replied:
You need to install the nodes, library, and launch files.

wxmerkt commented on src/polhemus_tf_broadcaster.cpp:
```diff
@@ -0,0 +1,292 @@
+	tf::Transform transform;
+
+	// Set translation
+	transform.setOrigin(tf::Vector3(0.0254*stations[i].x,
```
This is converting inches to cm I presume? Please add a comment explaining "magic" conversions.

wxmerkt commented on src/polhemus_tf_broadcaster.cpp:
```diff
@@ -0,0 +1,292 @@
+
+	// Set orientation
+	tf::Quaternion q;
+	q.setRPY(0.0174533*stations[i].ro,
```
Likewise, please add what conversion this is - and use quaternions again unless you have a specific reason other than debug for switching to Euler angles.

wxmerkt reviewed (changes requested):
General comments:

1. Please [squash your commits](https://davidwalsh.name/squash-commits-git) down to a handful of independently working commits with clear commit messages.
2. Consider using tf2 instead of tf - cf [1](https://answers.ros.org/question/11242/what-is-the-difference-between-tf-and-tf2/) and [2](http://wiki.ros.org/tf2/Migration).
3. Please revert to using quaternions instead of Euler angles unless there is a more important reason than debugging for the change - Euler angles have disadvantages (singularities and several rotations among others).
4. When reworking files use move/edit so we can track changes, e.g. files here show up as deleted and recreated which makes it incredibly hard to review (perhaps because they were also reformatted in the same go).
5. More comments inline

VladimirIvan commented on launch/start.launch:
```diff
@@ -0,0 +1,39 @@
+
+  <param name="station_8_attached" value="false" type="bool"/>
+  <param name="station_8_broadcast" value="false" type="bool"/>
+
```
Having just one flag for not/connected should be enough. Additional frames can just be ignored. Remove the broadcast flag and assume its always true.

VladimirIvan commented on src/polhemus_tf_broadcaster.cpp:
```diff
@@ -0,0 +1,292 @@
+	// Set translation
+	transform.setOrigin(tf::Vector3(0.0254*stations[i].x,
+					0.0254*stations[i].y,
+					0.0254*stations[i].z));
```
What is the magic constant `0.0254`?

VladimirIvan commented on src/polhemus_tf_broadcaster.cpp:
```diff
@@ -0,0 +1,292 @@
+  liberty_clear_input(handle); //right now, we just ignore the answer
+
+  // (***)
+  euler_station_t *stations = (euler_station_t*) (malloc(sizeof(euler_station_t) * nstations_found));
```
Use c++ ```new```/```delete``` instead of c ```malloc```/```free```.

VladimirIvan replied:
Even better, use a ```std::vector<euler_station_t>```

cmower replied:
@VladimirIvan Would using `std::vector<station_t>` require changes to `liberty.cpp` and `liberty.h`?

wxmerkt replied:
No, you just create the vector and ``push_back`` any ``station_t`` handle you create. No need to pre-allocate the size of a STL container

wxmerkt replied:
Actually, you were correct. Don't use a vector because it would require that change to liberty. Liberty appears to require a traditional C array.

cmower replied:
@VladimirIvan and @wxmerkt if I understand correctly, we have to keep `malloc`/`free` rather than c++ `new`/`delete` because the polhemus uses C arrays? If so, I think the driver is complete.

VladimirIvan commented on src/polhemus_tf_broadcaster.cpp:
```diff
@@ -0,0 +1,292 @@
+  // Shutdown
+  liberty_send(handle, (char *)"p"); // stop continuous mode
+  usb_close(handle);
+  free(stations);
```
```delete```

VladimirIvan reviewed (changes requested):
> We now extract Euler angles from polhemus using the euler_station_t struct in protocol.h rather than station_t that uses quaternions. More intuitive and easier to debug.
The euler angles are not used anywhere apart from conversion to a quaternion. Why use them then? There is nothing to debug.

Some of the renaming and removing can be squashed to one commit. Try to group changes a bit more in the future. E.g. it is not clear from the commit why 52498fe is not needed anymore.

wxmerkt commented:
We just had a chat and will do the squashing just before merging after all comments have been addressed

wxmerkt commented on src/polhemus_tf_broadcaster.cpp:
```diff
@@ -258,26 +259,27 @@ int main(int argc, char** argv) {
+	transformStamped.child_frame = "polhemus_station_" + std::to_string(s);
+	
+	// Set translation (inches)
+	transformStamped.transform.translation.x = stations[i].x;
```
ROS uses meters. So the conversion was okay but you need to document why and not just use a magic number

wxmerkt commented on src/polhemus_tf_broadcaster.cpp:
```diff
@@ -255,32 +224,31 @@ int main(int argc, char** argv) {
+      transformStamped.transform.translation.z = stations[i].z;
+
+      // Set rotation
+      transformStamped.transform.rotation.x = stations[i].quaternion[3];
```
This looks suspiciously wrong. Is the convention in Polhemus really y,z,w,x? It should be either w,x,y,z (bot-frames convention for instance) or x,y,z,w (ROS convention). Please check the docs!

cmower replied:
@wxmerkt here i am extracting (w,x,y,z), i.e. (q[3], q[0], q[1], q[2]). 

The docs for the polhemus liberty that say the order should be (q[3], q[0], q[1], q[2]) (as in the code), see page 21 [here](http://polhemus.com/_assets/img/LIBERTY_User_Manual_URM03PH156-H.pdf). 

However, we have a problem using the following. 
```
// Set rotation
transformStamped.transform.rotation.x = stations[i].quaternion[3];
transformStamped.transform.rotation.y = stations[i].quaternion[0];
transformStamped.transform.rotation.z = stations[i].quaternion[1];
transformStamped.transform.rotation.w = stations[i].quaternion[2];
```
Rotation about the x-axis works as intended but about the other two axes the direction shown in RVIZ is in reverse. I have solved the problem (see recent commits) by changing the sign of q[0] and q[1], as follows.

```
// Set rotation
transformStamped.transform.rotation.x = stations[i].quaternion[3];
transformStamped.transform.rotation.y = -stations[i].quaternion[0];
transformStamped.transform.rotation.z = -stations[i].quaternion[1];
transformStamped.transform.rotation.w = stations[i].quaternion[2];
```
I currently have no idea why this happens. Could this be a problem in either `protocol.h` or `liberty.cpp`/`liberty.h`?

wxmerkt replied:
Yeah that's what I meant - your order is wrong. Check what you are _assigning it to_. Your assigned order is xyzw (in the transformStamped) but you are assigning wxyz. Change the name of the variables you are assigning to and it should work.
And *never* change the sign of a quaternion (it's normalised).
(i.e. the problem is in your code and not in liberty)

wxmerkt replied:
@cmower Seeing that the updated hacked solution again involves flipping signs in components of the quaternion, I recommend having a look at how quaternions work [and why flipping signs is a bad idea, e.g. RSS lecture slides, wikipedia]. The issue is not in the order in which you are getting it from the liberty quaternion array (3,0,1,2 - WXYZ) but where you are assigning it to on the left hand-side (xyzw). Just changing the ".x/.y/.z/.w" on the left hand side (transformStamped) to the correct order will solve your problem. Happy to stop by 1.30 after lunch for a face to face conversation

cmower committed:
Moved to tf2 publishing

wxmerkt commented on README.md:
```diff
@@ -0,0 +1,43 @@
+
+This package publishes coordinate transforms for the Polhemus Liberty sensors (stations) using the [`tf2` package](http://wiki.ros.org/tf2).
+
+The user manual for the Polhemus Liberty can be found [here](http://polhemus.com/_assets/img/LIBERTY_User_Manual_URM03PH156-H.pdf).
```
I've also downloaded a backup and put it on ``BigNAS:home/Documents``

wxmerkt committed:
Fix CMakeLists.txt, add libusb-dev/fxload to dependencies, update README
- Correct export of library to catkin_package
- Install targets
- Add libusb-dev and fxload to package.xml dependency list
- CMake formatting
- Update package.xml

wxmerkt committed:
Re-add ability to Ctrl-C application

wxmerkt committed:
CMakeLists: More minimal way of finding libusb-dev

wxmerkt committed:
polhemus_tf_broadcaster: fix tf timestamp; rate limiting; use delete[]

wxmerkt reviewed (approved):
- Squashed, made compact and fixed a few bugs (e.g. exiting)
- Verified no memory leak present
- Tested with hardware device

wxmerkt commented:
@VladimirIvan This is all okay and good to go from my side. Went over the code and tested it with hardware, made it more minimal and fixed it in a few places. Ready to merge?

All in all great job @cmower, works nicely with ``tf`` :-).

There's now also a copy of the user manual on BigNAS, just in case the original disappears from the internet.
